informational/req-res-reliability.md

```diff
@@ -0,0 +1,89 @@
+---
+title: REQ-RES-RELIABILITY
+name: Request-response protocols reliability
+category: Best Current Practice
+tags: [informational]
+editor: Oleksandr Kozlov <oleksandr@status.im>
+contributors:
+- Prem Chaitanya Prathi <prem@status.im>
+- Danish Arora <danish@status.im>
+---
+
+## Abstract
+This RFC describes set of instructions used across different [WAKU2](https://github.com/vacp2p/rfc-index/blob/7b443c1aab627894e3f22f5adfbb93f4c4eac4f6/waku/standards/core/10/waku2.md) implementations for improved reliability during usage of request-response protocols by a light node:
+- [WAKU2-LIGHTPUSH](https://github.com/vacp2p/rfc-index/blob/7b443c1aab627894e3f22f5adfbb93f4c4eac4f6/waku/standards/core/19/lightpush.md) - is used for sending messages;
+- [WAKU2-FILTER](https://github.com/vacp2p/rfc-index/blob/7b443c1aab627894e3f22f5adfbb93f4c4eac4f6/waku/standards/core/12/filter.md) - is used for receiving messages;
+
+### Definitions
+- Service node - provides services to other nodes such as relaying messages send by LightPush to the network or broadcasts messages from the network through Filter, usually serves responses;
+- Light node - connects to and uses one or more service nodes via LightPush and/or Filter protocols, usually sends requests;
+
+## Motivation
+
+Specifications of the mentioned protocols do not define some of the real world use cases that are often observed in unreliable network environment from the perspective of light nodes that are consumers of LightPush and/or Filter protocols.
+Such use cases can be: recovery from offline state, decrease rate of missed messages, increase probability of messages being broadcasted within the network.
+
+## Suggestions
+
+### Node health
+
+Node health is a metric meant to determine how reliable a light node is.
+We consider this reliability to be dependant on amount of simultaneous connections to responsive service nodes.
+Unfortunately the more connections light node establishes - the more bandwidth is consumed.
+To address this we suggest following metrics:
+- unhealthy - no connections to service nodes are available regardless of protocol;
+- minimally healthy:
+  - Filter has one service node connection;
+  - LightPush protocol has one service node connection;
+- sufficiently healthy:
+  - Filter has at least 2 connections available to service nodes;
+  - LightPush has at least 2 connections available to service nodes;
+
+### Peers and connection management
+
+#### Pool of reliable service nodes
+Light node should maintain a pool of reliable service nodes for each protocol.
+In case service node fails to serve protocol request from a light node 3 times - light node should drop connection to it and a new service node should be connected and added to the pool instead.
```

Comment:
> 3 times seems high.

it can be, so with default `ping` interval it can be 3 minutes, and for `lightpush` it can be whatever it will take in order for user to send 3 requests 

adding better definition for light push and filter specifically

covered `a, b` as part of definition of failure
as for `c` I am not sure, I would consider it part of another strategy, will add a note to `Retry missing messages` subsection

 > If all remote nodes, then "dropping connection to it and using a new service node" is not really useful

I guess application should identify it is offline and/or their shard is not functioning

> If a single remote node is unreachable, then trying 3 times seems like a waste of time.

to me it seems as default assumption we do - service node fails 
but there are different kinds of failures: for Filter it is straightforward, probably, but for LightPush it might be different error that happened - I added a note saying that error code should be considered to make a decision to retry or drop.

Reply:
> 3 times seems high.

I am lowering it to 2 times and and setting in code 30 seconds for recurring pings
wdyt @danisharora099 ?